### ISSUE TITLE
Need way to specify `h5cc` explicitly

### ISSUE THREAD

**Issue by david-macmahon** [MydonSolutions/bfr5c99#4]
I encountered errors building because `meson` found the `h5cc` script in the miniconda environment from which `meson` was running rather than the correct file (`/usr/local/bin/h5pcc`) for the system I was building on.  This incorrect `h5cc` file included a compiler flag (`-fno-plt`) that was not compatible with the compiler version on that system leading to build errors.

The `README.md` and `meson --help` and a Google search (or two) didn't provide any clues as to how to specify the `h5cc` script explicitly.  If, as it seems, this cannot be done currently, then please add a way to do it.  `README.md` should mention how to specify this explicitly in the unlikely event the wrong one gets found by default.

**Comment by eli-schwartz** [comment]
> any clues as to how to specify the `h5cc` script explicitly.

https://mesonbuild.com/Machine-files.html#binaries

If you look at `builddir/meson-logs/meson-log.txt` you should see the following debug log:
```
h5cc binary missing from cross or native file, or env var undefined.
Trying a default h5cc fallback at h5cc
```

So, you can create a native file with a binaries section pointing to the h5cc absolute location. (Env vars will not work, we don't include that as one of the special programs with a tracked env var, like CC/CXX/PKG_CONFIG.)

BTW the most advantageous way to handle this is by using the pkg-config files provided by hdf5, but last time I checked, the cmake build of hdf5 is a bit... broken... and the autotools build didn't install pkg-config files. It's a bit of a mess...

If you could use pkg-config, then that's a lot easier to consistently override via $PKG_CONFIG_PATH of course.

**Comment by david-macmahon** [comment]
Yeah, I guess I'm asking for `h5cc` to be treated as a special program with a tracked env var.

Following the `meson` docs, I created a `native` file:

```
$ cat $HOME/.local/share/meson/native
[binaries]
h5cc = '/usr/local/bin/h5pcc'
```

This is where the `meson` docs say it should go, but it wasn't picked up automatically.  I could get it to work by running:

```
meson setup --native=$HOME/.local/share/meson/native build
```

While it's nice to have *a* solution, it would be nicer to have a *convenient* solution such a being able to specify this via an env var (i.e. without creating external config files).

**Comment by eli-schwartz** [comment]
> This is where the `meson` docs say it should go, but it wasn't picked up automatically.

My reading of the documentation is that you can create a native file e.g. `hdf.ini` inside the *directory* `$HOME/.local/share/meson/native/` and then run `meson setup --native=hdf.ini` without specifying a full path.

There are no default native files, currently. There's a proposal to have `siteconfig.ini` automatically loaded if it exists.



> While it's nice to have _a_ solution, it would be nicer to have a _convenient_ solution such a being able to specify this via an env var (i.e. without creating external config files).

I agree, this has been something that left me a bit dissatisfied in the past too. There's some work on creating default native files based on the environment (`meson env2mfile`, in the future we may automatically run this during setup).

**Comment by david-macmahon** [comment]
Ah, yes, I missed that detail, thanks for clarifying.  Is it really that convoluted/clunky to add support for another env var?  Meson is pretty new to me, but so far it seems overly opinionated IMHO...

**Comment by eli-schwartz** [comment]
It's not especially complicated -- there's a table of `'tool': 'ENV_VAR'` here: https://github.com/mesonbuild/meson/blob/5f3c7125cc4f9dadd084dee3be5fd84458ae74c8/mesonbuild/envconfig.py#L119

**Comment by david-macmahon** [comment]
Can that be added to at runtime from within a project's meson config file or would this have to be a PR for meson itself?

**Comment by eli-schwartz** [comment]
The latter. The h5cc tool is looked up via code built into Meson, anyway. https://mesonbuild.com/Dependencies.html#hdf5 see the config-tool method.

**Comment by david-macmahon** [comment]
Hmm, interesting, it seems like meson strives to be omniscient, but this is veering off topic into the philosophical, so I'll stop talking/whinging now :)

**Comment by radonnachie** [comment]
Coming from [this](https://mesonbuild.com/Dependencies.html#system), perhaps using [`compilerfind`](https://mesonbuild.com/Reference-manual_returned_compiler.html#compilerfind_library when a particular argument is supplied could satisfy this... 

but I think there is a disconnect between defining a custom hdf5 dependency and specifying the h5cc to use for that dependency.

---------------------
```
dep_lst = []

if get_option('h5cc') != ''
  cc = meson.get_compiler('c')
  dep_lst += [
    cc.find_library(
      'hdf5',
      dirs: [get_option('h5cc')],
      has_headers: ['hdf5.h'],
      required: true
    )
  ]
endif
if get_option('h5cc') == ''
  # auto find hdf5
  dep_lst += [dependency('hdf5', language:'c')]
endif
```
